tools/awps-tunnel/client/src/components/api/Parameters.tsx

```diff
@@ -0,0 +1,313 @@
+		setRequestParameterInputs(prev => ({
+			...prev, [name]: e.target.value
+		}))
+	}
+	
+	useEffect(() => {
+		let temp_parameter_input: { [name: string]: string } = {}
+		parameters.filter(p => p.in === "query" || p.in === "path").forEach((p) => {
+			if (p.name !== "api-version" && p.name !== "hub") {
+				temp_parameter_input[p.name] = ""
+			}
+		})
+		setRequestParameterInputs(temp_parameter_input);
+		if (example.groupsToAdd || example.groupsToRemove || example.message) {
+			setBodyParameterInputs(example.groupsToAdd || example.groupsToRemove || {message: example.message});
+		}
+		
+	}, [parameters, example]);
+	
+	useEffect(() => {
+		let temp_parameter: { parameter: Parameter, inputComponent: React.JSX.Element }[] = [];
+		parameters.filter(p => p.in === "query" || p.in === "path").forEach((p) => {
+			temp_parameter.push({
+				parameter: p,
+				inputComponent: <Input value={requestParameterInputs[p.name] || ""}
+				                       onChange={e => handleParameterOnChange(e, p.name)}/>
+			})
+		})
+		setRequestParameters(temp_parameter);
+		setRequestBody(parameters.filter(p => p.in === "body"));
+	}, [parameters, requestParameterInputs]);
+	
+	useEffect(() => {
+		if (requestBody.length > 0 && requestBody[0].schema) {
+			if (requestBody[0].schema.type === "string") {
+				const bodyParameter: Parameter = requestBody[0];
+				setBodySchema({name: bodyParameter.name, type: bodyParameter.schema?.type} as Definition)
+			} else {
+				const ref: string = requestBody[0].schema.$ref;
+				const path: string[] = ref.split('/');
+				const defName: string = path[path.length - 1];
+				if (defName === "AddToGroupsRequest") {
+					setBodySchema(restapiSpec.definitions["AddToGroupsRequest"]);
+				} else if (defName === "RemoveFromGroupsRequest") {
+					setBodySchema(restapiSpec.definitions["RemoveFromGroupsRequest"]);
+				}
+			}
+		} else {
+			setBodySchema(undefined);
+		}
+	}, [requestBody]);
+	
+	useEffect(() => {
+		let newPath = path;
+		if (data.hub) {
+			newPath = newPath.replace('{hub}', data.hub);
+		}
+		let query: string = "";
+		requestParameters.forEach(({parameter}): void => {
+			if (parameter.in === "path") {
+				newPath = newPath.replace(`{${parameter.name}}`, requestParameterInputs[parameter.name])
+			} else {
+				if (requestParameterInputs[parameter.name]) {
+					query += `${parameter.name}=${requestParameterInputs[parameter.name]}&`
+				}
+			}
+		})
+		setUrl(`${data.endpoint.slice(0, -1)}${newPath}?${query}api-version=${restapiSpec.info.version}`);
+	}, [path, data.hub, data.endpoint, requestParameterInputs, requestParameters]);
+	
+	useEffect(() => {
+		async function getToken() {
+			// if no server connected, typeof token if object
+			const token: any = await dataFetcher.invoke("getRestApiToken", url);
+			if (typeof (token) === "string") {
+				setToken(token);
+			} else {
+				setToken("please connect to server to get the token")
```

Comment:
token is used as error message? when error the error message is copied to clipboard? Use another property instead?

Reply:
This may occur when user directly start the project in cient folder by `npm start` since without starting a server, the client can't get an accessKey or connectionString to send http request.

Reply:
if the project is started by client instead of server, the `token` will be an object instead of a string